src/routes/auth/auth.service.ts

```diff
@@ -0,0 +1,156 @@
+import { CacheRouter } from '@/datasources/cache/cache.router';
+import { IConfigurationService } from '@/config/configuration.service.interface';
+import { SiweMessage } from '@/domain/auth/entities/siwe-message.entity';
+
+@Injectable()
+export class AuthService {
+  static readonly AUTH_TOKEN_TOKEN_TYPE = 'Bearer';
+
+  private readonly nonceTtlInSeconds: number;
+
+  constructor(
+    @Inject(IConfigurationService)
+    private readonly configurationService: IConfigurationService,
+    @Inject(IAuthRepository)
+    private readonly authRepository: IAuthRepository,
+    @Inject(IJwtService)
+    private readonly jwtService: IJwtService,
+    @Inject(CacheService) private readonly cacheService: ICacheService,
+  ) {
+    this.nonceTtlInSeconds = this.configurationService.getOrThrow(
+      'auth.nonceTtlSeconds',
+    );
+  }
+
+  /**
+   * Generates a unique nonce and stores it in cache for later verification.
+   *
+   * @returns nonce - unique string to be signed
+   */
+  async getNonce(): Promise<{
+    nonce: string;
+  }> {
+    const nonce = this.authRepository.generateNonce();
+
+    // Store nonce for reference to verify/prevent replay attacks
+    const cacheDir = CacheRouter.getAuthNonceCacheDir(nonce);
+    await this.cacheService.set(cacheDir, nonce, this.nonceTtlInSeconds);
+
+    return {
+      nonce,
+    };
+  }
+
+  /**
+   * Verifies the validity of a signed message and returns an access token:
+   *
+   * 1. Ensure the message itself has not expired.
+   * 2. Ensure the nonce was generated by us/is not a replay attack.
+   * 3. Verify the signature of the message.
+   * 4. Return an access token if all checks pass.
+   *
+   * @param args.request - Express request object.
+   * @param args.verifyAuthMessageDto - DTO containing the message and signature to verify.
+   *
+   * The following adhere to JWT standard {@link https://datatracker.ietf.org/doc/html/rfc7519}
+   *
+   * @returns accessToken - JWT access token
+   * @returns tokenType - token type ('Bearer') to be used in the `Authorization` header
+   * @returns notBefore - expoch from when token is valid (if applicable, otherwise null)
+   * @returns expiresIn - time in seconds until the token expires (if applicable, otherwise null)
+   */
+  async verify(args: {
+    request: Request;
+    verifyAuthMessageDto: VerifyAuthMessageDto;
+  }): Promise<{
+    accessToken: string;
+    tokenType: string;
+    notBefore: number | null;
+    expiresIn: number | null;
+  }> {
+    const isAuthorized = await this.isAuthorized(
+      args.verifyAuthMessageDto,
+    ).catch(() => false);
+
+    if (!isAuthorized) {
+      throw new UnauthorizedException();
+    }
+
+    const { message } = args.verifyAuthMessageDto;
+
+    const dateWhenTokenIsValid = message.notBefore
+      ? new Date(message.notBefore)
+      : null;
+    const dateWhenTokenExpires = message.expirationTime
+      ? new Date(message.expirationTime)
+      : null;
```

Comment:
Can't we get the `message` and respective field coercions with Zod? The `message` is an `hexSchema` so how would it have the properties `notBefore` and `expirationTime`?

Reply:
> Can't we get the `message` and respective field coercions with Zod? The `message` is an `hexSchema` so how would it have the properties `notBefore` and `expirationTime`?

`message` is [an object](https://github.com/safe-global/safe-client-gateway/blob/main/src/domain/auth/entities/siwe-message.entity.ts#L12). It does not coerce `notBefore` or `expirationTime` as the field values would have been signed _as is_ by the user.

We receive the values and then convert them to [the would-have-been-signed message](https://github.com/safe-global/safe-client-gateway/blob/main/src/datasources/auth-api/siwe-api.service.ts#L29) for comparison. If we coerced the values, we would have to change them back to a string when creating said message. Would it be preferable to do so?

Note: I understand the logic flow is quite complex and I intend to documen this.

Reply:
I see. I might have actually mixed it up with another schema because, `message` is not an `HexSchema`. Sorry for the confusion!